packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/useHandleUpload.ts

```diff
@@ -129,60 +152,148 @@ export function useHandleUpload({
+
+    return {
+      task: pendingTask,
+      result,
+    };
+  };
+
+  const handleUpload = async () => {
+    let currentIndex = 0;
+
+    // Update tasks to be in QUEUED state
+    // the 'cancel' property will update the the state so the task
+    // has status 'CANCELED' and its 'cancel' property is undefined
     setTasks((prevTasks) =>
-      prevTasks.map(({ key, data, progress, size }) => {
-        const { bucket: bucketName, credentialsProvider, region } = getConfig();
-        const input: UploadDataWithPathInput = {
-          path: `${prefix}${key}`,
-          data,
-          options: {
-            bucket: { bucketName, region },
-            locationCredentialsProvider: credentialsProvider,
-            onProgress: ({ totalBytes, transferredBytes }) => {
-              const progress = totalBytes ? transferredBytes / totalBytes : 0;
-              const nextTask: TaskUpdate = {
-                key,
-                progress,
-                ...(totalBytes === transferredBytes
-                  ? { cancel: undefined }
-                  : undefined),
-              };
-              setTasks((curr) => updateTasks(curr, nextTask));
-            },
-            preventOverwrite,
-          },
-        };
+      prevTasks.map((task) => ({
+        ...task,
+        status: 'QUEUED',
+        cancel: () =>
+          setTasks((prevTasks) =>
+            updateTasks(prevTasks, {
+              key: task.key,
+              status: 'CANCELED',
+              cancel: undefined,
+            })
+          ),
+      }))
+    );
 
-        const { cancel, result } = uploadData(input);
+    const uploadNext = async () => {
+      const currentTasks = tasksRef.current;
 
-        result
-          .then(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'COMPLETE' }));
+      if (currentIndex >= currentTasks.length) {
+        return;
+      }
+
+      const queuedTask = currentTasks[currentIndex];
+
+      if (queuedTask.status === 'CANCELED') {
+        currentIndex++;
+
+        uploadNext();
+
+        return;
+      }
+
+      const pendingTask = processUpload(queuedTask);
+
+      currentIndex++;
+
+      const { task, result } = pendingTask;
+      const { key } = task;
+
+      try {
+        await result;
+
+        setTasks((prevTasks) =>
+          updateTasks(prevTasks, { key, status: 'COMPLETE' })
+        );
+      } catch (error) {
+        setTasks((prevTasks) =>
+          updateTasks(prevTasks, {
+            key,
+            status: isCancelError(error) ? 'CANCELED' : 'FAILED',
+            cancel: undefined,
           })
-          .catch(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'FAILED' }));
-          });
+        );
+      } finally {
+        uploadNext();
+      }
+    };
 
-        const handleCancel = () => {
-          cancel();
+    // Start the initial set of uploads
+    const initialUploads = Array(batchSize)
```

Comment:
As discussed offline, this PR should cover a single operation at a time to get the base approach correctly before expanding to multiple concurrent operations